server/src/main/java/org/apache/cloudstack/agent/lb/IndirectAgentLBServiceImpl.java

```diff
@@ -136,17 +146,54 @@ public int compare(Long x, Long y) {
         }
         final List <Host> agentBasedHosts = new ArrayList<>();
         for (final Host host : allHosts) {
-            if (host == null || host.getResourceState() == ResourceState.Creating || host.getResourceState() == ResourceState.Error) {
-                continue;
+            conditionallyAddHost(agentBasedHosts, host);
+        }
+        return agentBasedHosts;
+    }
+
+    private void conditionallyAddHost(List<Host> agentBasedHosts, Host host) {
+        if (host == null) {
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("trying to add no host to a list");
             }
-            if (host.getType() == Host.Type.Routing || host.getType() == Host.Type.ConsoleProxy || host.getType() == Host.Type.SecondaryStorage || host.getType() == Host.Type.SecondaryStorageVM) {
-                if (host.getHypervisorType() != null && host.getHypervisorType() != Hypervisor.HypervisorType.KVM && host.getHypervisorType() != Hypervisor.HypervisorType.LXC) {
-                    continue;
-                }
-                agentBasedHosts.add(host);
+            return;
+        }
+
+        ResourceState[] allowedStates = new ResourceState[]{
+                ResourceState.Enabled,
+                ResourceState.Maintenance,
+                ResourceState.Disabled,
+                ResourceState.ErrorInMaintenance,
+                ResourceState.PrepareForMaintenance
+        };
+        // so the remaining ResourceState[] disallowedStates = new ResourceState[]{ResourceState.Creating, ResourceState.Error};
+        if (binarySearch(allowedStates, host.getResourceState()) < 0) {
```

Comment:
Why not use `list.contains()`?

Reply:
seems like less cpu power/more code?

Reply:
@rhtyd ?

Reply:
@DaanHoogland premature optimisation is... nope there won't be a lot of performance gain. If you have to you can use `Collections.binarySearch()`

Reply:
ok, it felt quite natural to me, but will use an EnumSet.contains().